compiler/plc_source/src/source_location.rs

```diff
@@ -0,0 +1,441 @@
+use std::{
+    fmt::{Debug, Formatter},
+    ops::Range,
+};
+
+use crate::{SourceCode, SourceContainer};
+
+#[derive(Clone, Default)]
+pub struct SourceLocationFactory {
+    file: Option<&'static str>,
+    newlines: NewLines,
+}
+
+impl SourceLocationFactory {
+    /// constructs a SourceRangeFactory used for internally generated code (e.g. builtins)
+    pub fn internal(src: &str) -> Self {
+        SourceLocationFactory { file: None, newlines: NewLines::build(src) }
+    }
+
+    /// constructs a SourceRangeFactory used to construct SourceRanes that point into the given source
+    pub fn for_source(source_code: &SourceCode) -> Self {
+        SourceLocationFactory {
+            file: Some(source_code.get_location_str()),
+            newlines: NewLines::build(&source_code.source),
+        }
+    }
+
+    /// creates a new SourceRange using the factory's file_name
+    pub fn create_range(&self, range: core::ops::Range<usize>) -> SourceLocation {
+        let start = TextLocation::from_offset(range.start, &self.newlines);
+        let end = TextLocation::from_offset(range.end, &self.newlines);
+        SourceLocation { span: CodeSpan::Range(start..end), file: self.file }
+    }
+
+    pub fn create_block_location(&self, local_id: usize, execution_order: Option<usize>) -> SourceLocation {
+        SourceLocation {
+            span: CodeSpan::Block { local_id, execution_order, inner_range: None },
+            file: self.file,
+        }
+    }
+
+    pub fn create_file_only_location(&self) -> SourceLocation {
+        SourceLocation { span: CodeSpan::None, file: self.file }
+    }
+}
+
+/// The location of a certain element in a text file
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct TextLocation {
+    /// Line in the source code where this location points to
+    line: usize,
+    /// Column in the sourcecode where this location points o
+    column: usize,
+    /// Raw offset to this location from the start of the file
+    offset: usize,
+}
+
+impl TextLocation {
+    pub fn new(line: usize, column: usize, offset: usize) -> Self {
+        TextLocation { line, column, offset }
+    }
+
+    pub fn from_offset(offset: usize, newlines: &NewLines) -> Self {
+        let line = newlines.get_line_nr(offset);
+        let column = newlines.get_column(line, offset);
+        TextLocation { line, column, offset }
+    }
+}
+
+/// Represents the location of a code element in a source code
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum CodeSpan {
+    /// The location of a block in a diagram
+    Block { local_id: usize, execution_order: Option<usize>, inner_range: Option<Range<usize>> },
```

Comment:
finding the execution order here is a bit surprising to me. Is there a specific reason? or is it just practical?

Reply:
So the idea was to have the identifiable location for a block available.
This is for me either the execution order, or when that is missing the local id. The reasoning is that in PLC Developer we used execution orders for debugging. It also makes sense from a user perspective to see that block 5 has an issue, where 5 is he execution order. This however does not work for blocks with no execution order such as input variables. I'm open to leaving he exec id out here. This just means that the debug has to revert to local id, which might be strange to see in stack traces or the UI has to re-adjust the local-id to execution id